### ISSUE TITLE
Rust building and testing fails while bootstrapping it in poky

### ISSUE THREAD

**Issue by Yashinde145** [rust-lang/rust#115642]
<!--
Thank you for filing a regression report! 🐛 A regression is something that changed between versions of Rust but was not supposed to.

Please provide a short summary of the regression, along with any information you feel is relevant to replicate it.
-->

Getting error while building & testing rust 1.72 via the `src/bootstrap/bootstrap.py`  file in poky (cmd- `bitbake rust` and then `python3 src/bootstrap/bootstrap.py test test_suites_names --target x86_64-poky-linux-gnu`)in poky environment. Until rust 1.70 I was able to build & test the rust sources successfully. 
But after updating to rust 1.72, there's an assertion failure causing the problem. The rust build fails as follows-

```rust
Note that not all variants of mir-opt tests are going to be blessed, as no mapping between a 
32bit and a 64bit target was found for x86_64-poky-linux-gnu.
You can add that mapping by changing MIR_OPT_BLESS_TARGET_MAPPING in src/bootstrap/test.rs
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `x86_64-unknown-linux-gnu`,
 right: `x86_64-poky-linux-gnu`: Cannot obtain compiler for non-native build triple at stage 0', compile.rs:1474:13
 ```

The code at compile.rs:1474

```
1470     fn run(self, builder: &Builder<'_>) -> Compiler {
1471         let target_compiler = self.target_compiler;
1472 
1473         if target_compiler.stage == 0 {
1474             assert_eq!(
1475                 builder.config.build, target_compiler.host,
1476                 "Cannot obtain compiler for non-native build triple at stage 0"
1477             );
1478             // The stage 0 compiler for the build triple is always pre-built.
1479             return target_compiler;
1480         }
```


Build terminates as follows-

```
test [mir-opt] tests/mir-opt/reference_prop.rs ... ok
test [mir-opt] tests/mir-opt/inline/polymorphic_recursion.rs ... ok
test result: ok. 249 passed; 0 failed; 13 ignored; 0 measured; 0 filtered out; finished in 1.39s

error: the option `Z` is only accepted on the nightly compiler
note: selecting a toolchain with `+toolchain` arguments require a rustup proxy; see <https://rust-lang.github.io/rustup/concepts/index.html>
help: consider switching to a nightly toolchain: `rustup default nightly`
note: for more information about Rust's stability policy, see <https://doc.rust-lang.org/book/appendix-07-nightly-rust.html#unstable-features>
thread 'main' panicked at 'failed to gather the target spec for i686-unknown-linux-gnu', synthetic_targets.rs:66:9
Build completed unsuccessfully in 0:05:27
```


### Version it worked on

<!--
Provide the most recent version this worked on, for example:

It most recently worked on: 
-->It most recently worked on Rust 1.70 and earlier versions
### Version with regression

<!--
Provide the version you are using that has the regression.
-->Rust 1.72


<!--
Did the compiler crash? If so, please provide a backtrace.
-->

### Backtrace
<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>

```
stack backtrace:                                                                                                                                                                                                
   0: rust_begin_unwind                                                                                                                                                                                         
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/std/src/panicking.rs:593:5                                                                                                              
   1: core::panicking::panic_fmt                                                                                                                                                                                
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/panicking.rs:67:14                                                                                                             
   2: core::panicking::assert_failed_inner                                                                                                                                                                      
   3: core::panicking::assert_failed                                                                                                                                                                            
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/panicking.rs:228:5                                                                                                             
   4: <bootstrap::compile::Assemble as bootstrap::builder::Step>::run                                                                                                                                           
             at ./src/bootstrap/compile.rs:1474:13                                                                                                                                                              
   5: bootstrap::builder::Builder::ensure                                                                                                                                                                       
             at ./src/bootstrap/builder.rs:2055:23                                                                                                                                                              
   6: bootstrap::builder::Builder::compiler                                                                                                                                                                     
             at ./src/bootstrap/builder.rs:943:9                                                                                                                                                                
   7: <bootstrap::test::Compiletest as bootstrap::builder::Step>::run                                                                                                                                           
             at ./src/bootstrap/test.rs:1462:24                                                                                                                                                                 
   8: bootstrap::builder::Builder::ensure                                                                                                                                                                       
             at ./src/bootstrap/builder.rs:2055:23                                                                                                                                                              
   9: <bootstrap::test::UiFullDeps as bootstrap::builder::Step>::run                                                                                                                                            
             at ./src/bootstrap/test.rs:1260:17                                                                                                                                                                 
  10: bootstrap::builder::Builder::ensure                                                                                                                                                                       
             at ./src/bootstrap/builder.rs:2055:23                                                                                                                                                              
  11: <bootstrap::test::UiFullDeps as bootstrap::builder::Step>::make_run                                                                                                                                       
             at ./src/bootstrap/test.rs:1256:17                                                                                                                                                                 
  12: bootstrap::builder::StepDescription::maybe_run                                                                                                                                                            
             at ./src/bootstrap/builder.rs:278:13                                                                                                                                                               
  13: bootstrap::builder::StepDescription::run                                                                                                                                                                  
             at ./src/bootstrap/builder.rs:315:21                                                                                                                                                               
  14: bootstrap::builder::Builder::run_step_descriptions                                                                                                                                                        
             at ./src/bootstrap/builder.rs:935:9                                                                                                                                                                
  15: bootstrap::builder::Builder::execute_cli                                                                                                                                                                  
             at ./src/bootstrap/builder.rs:916:9                                                                                                                                                                
  16: bootstrap::Build::build                                                                                                                                                                                   
             at ./src/bootstrap/lib.rs:697:17                                                                                                                                                                   
  17: bootstrap::main                                                                                                                                                                                           
             at ./src/bootstrap/bin/main.rs:74:5                                                                                                                                                                
  18: core::ops::function::FnOnce::call_once                                                                                                                                                                    
             at /rustc/8ede3aae28fe6e4d52b38157d7bfe0d3bceef225/library/core/src/ops/function.rs:250:5                                                                                                          
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.                                                                                                                         
Build completed unsuccessfully in 0:00:20


<!--
@rustbot modify labels: +regression-from-stable-to-{stable} -regression-untriaged

**Comment by Nilstrieb** [comment]
I believe `MIR_OPT_BLESS_TARGET_MAPPING` is unrelated to the issue you're seeing, but I'm not sure.

**Comment by Yashinde145** [comment]
Yeah, I guess the assertion failure is main cause. 
Why the two strings are different though? Until rust 1.70, I didn't faced this assertion failure(diff in `builder.config.build`  i.e left: `x86_64-unknown-linux-gnu`, and `target_compiler.host` i.e  right: `x86_64-poky-linux-gnu`). When did git blame for these lines, found that this code was present (from 5 years i.e 2018)even in the earlier rust versions.

**Comment by apiraino** [comment]
Adding useful labels for triaging 

@rustbot label -I-prioritize +T-bootstrap -needs-triage

**Comment by onur-ozkan** [comment]
I couldn't reproduce this problem from the e39976ff89f91b742916349859e8d877a4876783 HEAD.

The assertation passes as expected(on `./x test bootstrap --target=..), I also printed them out.
![image](https://github.com/rust-lang/rust/assets/39852038/805d2ff1-5b9f-481b-b09d-04fed9489c32)

The diff:
![image](https://github.com/rust-lang/rust/assets/39852038/2a37998d-fb3a-4937-8a1b-c2dbb29ca2ad)

Sorry for sharing the the outputs in the image format; the attempt was done inside of QEMU.

**Comment by Yashinde145** [comment]
My build shows following- 
Note- The targets are printed 3 times and the assertion fails in the 3rd call

```
Compiling clap v4.2.4
   Compiling clap_complete v4.2.2
    Finished dev [unoptimized] target(s) in 20.26s

builder.config.build = x86_64-unknown-linux-gnu
target_compiler.host = x86_64-unknown-linux-gnu
builder.config.build = x86_64-unknown-linux-gnu
target_compiler.host = x86_64-unknown-linux-gnu
builder.config.build = x86_64-unknown-linux-gnu
target_compiler.host = x86_64-poky-linux-gnu
```

Moreover, if I bypass the assertion check and then build, why do the compiler asks to switch to nightly channel instead of continuing to stable? I am wondering how does it gets the `Z` option during the build since I did not not provide any such options.

```
test [mir-opt] tests/mir-opt/reference_prop.rs ... ok
test [mir-opt] tests/mir-opt/inline/polymorphic_recursion.rs ... ok
test result: ok. 249 passed; 0 failed; 13 ignored; 0 measured; 0 filtered out; finished in 1.39s

error: the option `Z` is only accepted on the nightly compiler
note: selecting a toolchain with `+toolchain` arguments require a rustup proxy; see <https://rust-lang.github.io/rustup/concepts/index.html>
help: consider switching to a nightly toolchain: `rustup default nightly`
note: for more information about Rust's stability policy, see <https://doc.rust-lang.org/book/appendix-07-nightly-rust.html#unstable-features>
thread 'main' panicked at 'failed to gather the target spec for i686-unknown-linux-gnu', synthetic_targets.rs:66:9
Build completed unsuccessfully in 0:05:27
```

**Comment by onur-ozkan** [comment]
Do you have a custom `config.toml`?

**Comment by Yashinde145** [comment]
The `config.toml` generated during the rust build is as follows-

`build_dir` is the directory where I am building rust (using cmd `bitbake rust` in yocto poky sources)

```
[target.x86_64-poky-linux-gnu]                                                                                                                                                                                  
llvm-config = "build_dir/recipe-sysroot/usr/lib/llvm-rust/bin/llvm-config"                                                                         
cxx = "build_dir/wrapper/target-rust-cxx"                                                                                                          
cc = "build_dir/wrapper/target-rust-cc"                                                                                                            
linker = "build_dir/wrapper/target-rust-ccld"                                                                                                      
                                                                                                                                                                                                                
[target.x86_64-unknown-linux-gnu]                                                                                                                                                                               
llvm-config = "build_dir/recipe-sysroot-native/usr/lib/llvm-rust/bin/llvm-config"                                                                  
cxx = "build_dir/wrapper/build-rust-cxx"                                                                                                           
cc = "build_dir/wrapper/build-rust-cc"                                                                                                             
linker = "build_dir/wrapper/build-rust-ccld"                                                                                                       
                                                                                                                                                                                                                
[llvm]                                                                                                                                                                                                          
static-libstdcpp = false                                                                                                                                                                                        
                                                                                                                                                                                                                
[rust]                                                                                                                                                                                                          
rpath = true                                                                                                                                                                                                    
channel = "stable"                                                                                                                                                                                              
optimize = true                                                                                                                                                                                                 
verbose-tests = true                                                                                                                                                                                            
                                                                                                                                                                                                                
[build]                                                                                                                                                                                                         
submodules = false                                                                                                                                                                                              
docs = false                                                                                                                                                                                                    
rustc = "build_dir/rust-snapshot/bin/rustc"                                                                                                        
profiler = true                                                                                                                                                                                                 
cargo = "build_dir/rust-snapshot/bin/cargo"                                                                                                        
vendor = true                                                                                                                                                                                                   
target = ["x86_64-poky-linux-gnu"]                                                                                                                                                                              
host = ["x86_64-poky-linux-gnu"]                                                                                                                                                                                
build = "x86_64-unknown-linux-gnu"                                                                                                                                                                              
                                                                                                                                                                                                                
[install]                                                                                                                                                                                                       
prefix = "build_dir/image/usr"                                                                                                                     
bindir = "build_dir/image/usr/bin"                                                                                                                 
libdir = "build_dir/image/usr/lib"                                                                                                                 
datadir = "build_dir/image/usr/share"                                                                                                              
mandir = "build_dir/image/usr/share/man"
```

**Comment by onur-ozkan** [comment]
iirc `bitbake rust` will give you the stable 1.68 compiler. If so, please remove `build.cargo` and `build.rust` from your configuration, and use the beta compiler which will be downloaded from CI automatically during the bootstrapping.

**Comment by Yashinde145** [comment]
Actually, `bitbake rust` will give stable 1.70 compiler(on master branch). I am manually upgrading the version to 1.72 and then getting the error in bootstrapping build.

I didn't understand removing `build.cargo and build.rust` and using beta compiler from CI during bootstrapping. Can you enlighten me on it?

**Comment by onur-ozkan** [comment]
> Actually, bitbake rust will give stable 1.70 compiler(on master branch).

Still, this is wrong.


Remove the following lines:

```toml
rustc = "build_dir/rust-snapshot/bin/rustc"                                                                                                        
cargo = "build_dir/rust-snapshot/bin/cargo"                                                                                                        
```

from your configuration

**Comment by Yashinde145** [comment]
Ok, will check and get back.

**Comment by Yashinde145** [comment]
Thanks for the suggestion though, but I can't modify `config.toml`. The same `config.toml` worked with previous rust versions. Actually, the build compiler fetches some `-Z` (nightly build flag option) while building on the stable channel and it's recommended not to shift to nightly build. Is there any way I can find which flag causes this build failure and suppress that flag to proceed the build?

**Comment by onur-ozkan** [comment]
> @onur-ozkan, any updates?

Sorry, I somehow missed the notification.

> @onur-ozkan Thanks for the suggestion, but I can't modify config.toml. The same config.toml worked with previous rust versions. Actually, the build compiler fetches some -Z (nightly build flag option) while building on the stable channel and it's recommended not to shift to nightly build. Is there any way I can find which flag causes this build failure and suppress that flag to proceed the build?

Bootstrap requires the immediate previous version to compile the next version, allowing it to handle unstable features. So there's an issue with your configuration (rustc and cargo), and you should find a way to update your config file. I'm surprised you didn't encounter an error message like: `Unexpected rustc version: 1.70.0; we should use **** to build the source with ****.` It might be more helpful to receive this error message at the beginning of the bootstrap process, I suppose.

**Comment by Yashinde145** [comment]
@onur-ozkan 

Actually, I see that `-Zdoctest-in-workspace` a nightly flag is passed by default in Rust 1.72 release notes [here](https://github.com/rust-lang/rust/releases/tag/1.72.0#:~:text=Enable%20%2DZdoctest%2Din%2Dworkspace%20by%20default.%20When%20running%20each%20documentation%20test%2C%20the%20working%20directory%20is%20set%20to%20the%20root%20directory%20of%20the%20package%20the%20test%20belongs%20to.%20docs%20%2312221%20%2312288)

More info- [docs](https://doc.rust-lang.org/nightly/cargo/commands/cargo-test.html#working-directory-of-tests) https://github.com/rust-lang/cargo/pull/12221 https://github.com/rust-lang/cargo/pull/12288

I suspect that `-Zdoctest-in-workspace` is causing buid failure as seen in above comment here, https://github.com/rust-lang/rust/issues/115642#issuecomment-1718806226.

**Comment by onur-ozkan** [comment]
As I explained above, you cannot build upstream rustc with 1.70 anymore, even for non-poky environments. You should either not set rustc manually in the build configuration or set it to the proper version.

**Comment by onur-ozkan** [comment]
> As I explained above, you cannot build upstream rustc with 1.70 anymore, even for non-poky environments. You should either not set rustc manually in the build configuration or set it to the proper version.

Alternatively, if this doesn't help, you can build the 1.71 source with the 1.70 compiler (which you already have) and then use that compiler (which is version 1.71) to bootstrap version 1.72. You shouldn't make version jumps higher than one major version. We also have [this check here](https://github.com/rust-lang/rust/blob/5113ed28ea1451a13eae3a05dca0dbabfd56f587/src/bootstrap/src/core/config/config.rs#L2272-L2282) which is supposed to prevent that.

**Comment by Yashinde145** [comment]
Actually, rust has now been upgraded to 1.74.1 version in poky.
I had tested all the versions from 1.72 to 1.75(manually upgraded the rust version in my local poky source) and this error still persists.

> Alternatively, if this doesn't help, you can build the 1.71 source with the 1.70 compiler (which you already have) and then use that compiler (which is version 1.71) to bootstrap version 1.72. You shouldn't make version jumps higher than one major version. We also have [this check here](https://github.com/rust-lang/rust/blob/5113ed28ea1451a13eae3a05dca0dbabfd56f587/src/bootstrap/src/core/config/config.rs#L2272-L2282) which is supposed to prevent that.

The above approach would not be sustainable since poky is frequently rebased with latest rust release versions.
I suspect the following commits(in rust 1.72), related to setting up of target configs might cause these issues:

https://github.com/rust-lang/rust/commit/00cc815e57329bb78d6df886275b3372ee8991be
https://github.com/rust-lang/rust/commit/a56c829e747a6ee2e1a1d8d20472f3a5a4d95c79

Can you confirm this?

**Comment by onur-ozkan** [comment]
> I suspect the following commits(in rust 1.72), related to setting up of target configs might cause these issues:

> https://github.com/rust-lang/rust/commit/00cc815e57329bb78d6df886275b3372ee8991be
> https://github.com/rust-lang/rust/commit/a56c829e747a6ee2e1a1d8d20472f3a5a4d95c79

> Can you confirm this?

I don't think they are related.

> Actually, rust has now been upgraded to 1.74.1 version in poky.
I had tested all the versions from 1.72 to 1.75(manually upgraded the rust version in my local poky source) and this error still persists.

Then we have another problem.

We are using a nightly feature here:

https://github.com/rust-lang/rust/blob/8d39ec1825024f3014e1f847942ac5bbfcf055b0/src/bootstrap/src/core/build_steps/synthetic_targets.rs#L59-L62

so using `channel="stable"` could cause this failure, but only for `mir-opt` tests with `--bless` (`x test mir-opt --bless`) mode. But reading the issue title you get the same error with `x build` too I guess?

**Comment by Yashinde145** [comment]
> so using channel="stable" could cause this failure, but only for mir-opt tests with --bless (x test mir-opt --bless) mode. But reading the issue title you get the same error with x build too I guess?

The error is given out when the compiler is built internally during bootstrapping rust compiler by following command:
`python3 src/bootstrap/bootstrap.py test --target x86_64-poky-linux-gnu `

If `x build` command is given explicitly in terminal, it works fine.

**Comment by onur-ozkan** [comment]
> > so using channel="stable" could cause this failure, but only for mir-opt tests with --bless (x test mir-opt --bless) mode. But reading the issue title you get the same error with x build too I guess?
> 
> The error is given out when the compiler is built internally during bootstrapping rust compiler by following command: `python3 src/bootstrap/bootstrap.py test --target x86_64-poky-linux-gnu `
> 
> If `x build` command is given explicitly in terminal, it works fine.

Can you apply the following patch and try testing it again? It should fix the problem you are facing.

```diff
diff --git a/src/bootstrap/src/core/build_steps/test.rs b/src/bootstrap/src/core/build_steps/test.rs
index 92140b00da8..a9ec804e65d 100644
--- a/src/bootstrap/src/core/build_steps/test.rs
+++ b/src/bootstrap/src/core/build_steps/test.rs
@@ -1527,11 +1527,15 @@ fn run(self, builder: &Builder<'_>) {
             for target in targets {
                 run(target);
 
-                let panic_abort_target = builder.ensure(MirOptPanicAbortSyntheticTarget {
-                    compiler: self.compiler,
-                    base: target,
-                });
-                run(panic_abort_target);
+                if builder.config.channel == "dev" || builder.config.channel == "nightly" {
+                    let panic_abort_target = builder.ensure(MirOptPanicAbortSyntheticTarget {
+                        compiler: self.compiler,
+                        base: target,
+                    });
+                    run(panic_abort_target);
+                } else {
+                    builder.info(&format!("INFO: The panic=abort MIR-opt test was skipped because it cannot run on the {} channel, as it requires nightly features.\n", builder.config.channel));
+                }
             }
         } else {
             run(self.target);
```

**Comment by Yashinde145** [comment]
Thanks for the above patch but still the build fails

```
   Compiling serde_json v1.0.99
   Compiling build_helper v0.1.0 (/home/poky/build-st/tmp/work/core2-64-poky-linux/rust/1.74.1/rustc-1.74.1-src/src/tools/build_helper)
   Compiling rustfix v0.6.1
   Compiling compiletest v0.0.0 (/home/poky/build-st/tmp/work/core2-64-poky-linux/rust/1.74.1/rustc-1.74.1-src/src/tools/compiletest)
    Finished release [optimized] target(s) in 13.47s
Testing stage1 compiletest suite=ui mode=ui (x86_64-unknown-linux-gnu -> x86_64-poky-linux-gnu)
thread 'main' panicked at src/tools/compiletest/src/common.rs:519:30:
no entry found for key
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::option::expect_failed
   3: <compiletest::common::Config>::target_cfgs
   4: <compiletest::header::needs::CachedNeedsConditions>::load
   5: compiletest::make_tests
   6: compiletest::run_tests
   7: compiletest::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Build completed unsuccessfully in 0:03:59
```

src/tools/compiletest/src/common.rs:519 is as follows:

```
514 
515     fn get_current_target_config(
516         config: &Config,
517         targets: &HashMap<String, TargetCfg>,
518     ) -> TargetCfg {
519         let mut cfg = targets[&config.target].clone();
520 
521         // To get the target information for the current target, we take the target spec obtained
522         // from `--print=all-target-specs-json`, and then we enrich it with the information
523         // gathered from `--print=cfg --target=$target`.
524         //
525         // This is done because some parts of the target spec can be overridden with `-C` flags,
526         // which are respected for `--print=cfg` but not for `--print=all-target-specs-json`. The
527         // code below extracts them from `--print=cfg`: make sure to only override fields that can
528         // actually be changed with `-C` flags.
529         for config in
530             rustc_output(config, &["--print=cfg", "--target", &config.target]).trim().lines()
```

**Comment by onur-ozkan** [comment]
That seems to be related with https://github.com/rust-lang/rust/commit/a56c829e747a6ee2e1a1d8d20472f3a5a4d95c79 which you linked above. Will figure out the solution.

**Comment by Yashinde145** [comment]
> That seems to be related with https://github.com/rust-lang/rust/commit/a56c829e747a6ee2e1a1d8d20472f3a5a4d95c79 which you linked above. Will figure out the solution.

Yes, let me know if you need any more information.

According to this https://github.com/rust-lang/rust/commit/a56c829e747a6ee2e1a1d8d20472f3a5a4d95c79 and https://github.com/rust-lang/rust/commit/00cc815e57329bb78d6df886275b3372ee8991be commits, the standard rust supported target specifications are obtained by `--print=all-target-specs-json` and `--print=cfg --target=$target` flags.

And to handle custom target specs, which are not included in `--print=all-target-specs-json` like poky target triples(x86_64-poky-linux-gnu in this case) `--print=target-spec-json"` is passed on the `TARGET_TRIPLE.json i.e x86_64-poky-linux-gnu.json`  file (in this case) and the target specs are parsed from it. 

This should actually add the custom target specs in the build process but I wonder why it doesn't do so...

Note: I already exported `$ export RUST_TARGET_PATH=/home/build-st/tmp/work/core2-64-poky-linux/rust/1.74.1/rust-targets` path were the `x86_64-poky-linux-gnu.json` and `x86_64-unknown-linux-gnu.json` file are located.


```
$ cat  x86_64-poky-linux-gnu.json 
{
    "llvm-target": "x86_64-poky-linux-gnu",
    "data-layout": "e-m:e-i64:64-f80:128-n8:16:32:64-S128",
    "max-atomic-width": 64,
    "target-pointer-width": "64",
    "target-c-int-width": "32",
    "target-endian": "little",
    "arch": "x86_64",
    "os": "linux",
    "env": "gnu",
    "vendor": "unknown",
    "target-family": "unix",
    "linker": "x86_64-poky-linux-gcc",
    "cpu": "x86-64",
    "features": "+sse3",
    "dynamic-linking": true,
    "executables": true,
    "linker-is-gnu": true,
    "linker-flavor": "gcc",
    "has-rpath": true,
    "position-independent-executables": true,
    "panic-strategy": "unwind"
}
```

**Comment by onur-ozkan** [comment]
Could you please test the last two commits from https://github.com/onur-ozkan/rust/commits/panic-abort-mir-opt when you have the time? The first commit should resolve the error of using the nightly flag with a stable compiler, and the second one should resolve the custom target issue.

**Comment by Yashinde145** [comment]
Thanks.
I will test with the given changes and revert back in a day(it's eod here).

**Comment by Yashinde145** [comment]
> Could you please test the last two commits from https://github.com/onur-ozkan/rust/commits/panic-abort-mir-opt when you have the time? The first commit should resolve the error of using the nightly flag with a stable compiler, and the second one should resolve the custom target issue.


I tested with the two commits and the errors of  `using nightly flag and custom target specs` were not seen.
Testing was completed for the test suites like ui, run-pass-valgrind, coverage, mir-opt, codegen, assembly, incremental.
 
 But, after that while coping libs to qemu emulator(we run the test suite on qemu), following io::copy panic was observed:
 
 
```
 REMOTE copy libs to emulator (x86_64-poky-linux-gnu)                                                                          
 Building stage0 tool remote-test-server (x86_64-unknown-linux-gnu -> x86_64-poky-linux-gnu)
 
                        
 done pushing "/home/build-st/tmp/work/core2-64-poky-linux/rust/1.75.0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-poky-linux-gnu/lib/libcstr-d49e684941d617ce.so"                                   
 done pushing "/home/build-st/tmp/work/core2-64-poky-linux/rust/1.75.0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-poky-linux-gnu/lib/libzerovec_derive-52a1b6c914833752.so"                         
 done pushing "/home/build-st/tmp/work/core2-64-poky-linux/rust/1.75.0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-poky-linux-gnu/lib/librustc_fluent_macro-4b4331439c3bd942.so"
 
thread 'main' panicked at src/tools/remote-test-client/src/main.rs:353:5: 
io::copy(&mut file, dst) failed with Connection reset by peer (os error 104)                                                  
 stack backtrace:                                                                                                              
    0: rust_begin_unwind                                                                                                       
              at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/panicking.rs:597:5                            
    1: core::panicking::panic_fmt                                                                                              
              at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/panicking.rs:72:14                           
    2: remote_test_client::send                                                                                                
    3: remote_test_client::main                                                                                                
 note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.                                       
 Build completed unsuccessfully in 0:09:27
```

**Comment by onur-ozkan** [comment]
Can you try again? That seems like related with a TCP connection on `remote-test-client` (and we didn't change anything on remote-test tools since 2 years including `1.70.0` and `1.72.0`).

**Comment by Yashinde145** [comment]
Sure, I will check again and revert back.

**Comment by Yashinde145** [comment]
I tested 4 to 5 times and got the same error each time.

**Comment by onur-ozkan** [comment]
1.70.0 and 1.72.0 remote-test tools are identical. I am not entirely sure what could go wrong. Is the workflow you doing is same as you did for 1.70 (the one was working before) ?

**Comment by Yashinde145** [comment]
Yes, I am doing the same workflow in both cases. 
Note- I am testing it on rust 1.75 built in poky sources as mentioned in https://github.com/rust-lang/rust/issues/115642#issuecomment-1878339998.

**Comment by onur-ozkan** [comment]
>  thread 'main' panicked at src/tools/remote-test-client/src/main.rs:353:5: 
>  io::copy(&mut file, dst) failed with Connection reset by peer (os error 104)

Error says that the server is shutting down before generating/sending a response. Can you provide more details about the part of your workflow that uses remote-test tools?

**Comment by Yashinde145** [comment]
We are using `dropbear` for ssh server.

You can find the complete workflow here: 
https://git.openembedded.org/openembedded-core/tree/meta/lib/oeqa/selftest/cases/rust.py

Details for remote-test tools:
```
# wrap the execution with a qemu instance.
# Tests are run with 512 tasks in parallel to execute all tests very quickly
with runqemu("core-image-minimal", runqemuparams = "nographic", qemuparams = "-m 512") as qemu:
            # Copy remote-test-server to image through scp
            host_sys = get_bb_var("RUST_BUILD_SYS", "rust")
            ssh = SSHControl(ip=qemu.ip, logfile=qemu.sshlog, user="root")
            ssh.copy_to(builddir + "/build/" + host_sys + "/stage1-tools-bin/remote-test-server","~/")
            # Execute remote-test-server on image through background ssh
            command = '~/remote-test-server --bind 0.0.0.0:12345 -v'
            sshrun=subprocess.Popen(("ssh", '-o',  'UserKnownHostsFile=/dev/null', '-o',  'StrictHostKeyChecking=no', '-f', "root@%s" % qemu.ip, command), shell=False, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

**Comment by onur-ozkan** [comment]
Here we have an issue for the remote-test tool problem https://github.com/rust-lang/rust/issues/97669

It seems to occur randomly in QEMU environments from time to time. Retrying seems to fix the issue for each occurrence. I will try to identify the root cause in the coming week.

**Comment by onur-ozkan** [comment]
#119619 PR will close this issue upon merging, as we have a separate issue open for the other problem.

**Comment by Yashinde145** [comment]
@onur-ozkan , Thanks for the help to fix this issue.

> Here we have an issue for the remote-test tool problem https://github.com/rust-lang/rust/issues/97669
> It seems to occur randomly in QEMU environments from time to time. Retrying seems to fix the issue for each occurrence. I will try to identify the root cause in the coming week.

Yeah, I faced this panic failures randomly after the test suite testing is completed starting from rust 1.71 version

**Comment by Yashinde145** [comment]
Just remembered now,

How do I handle the assertion mentioned in https://github.com/rust-lang/rust/issues/115642#issue-1885966668?
Until now I was by-passing the assertion check with the following changes:
Should I continue with it?
```
diff --git a/src/bootstrap/src/core/build_steps/compile.rs b/src/bootstrap/src/core/build_steps/compile.rs                          
--- a/src/bootstrap/src/core/build_steps/compile.rs                                                                                 
+++ b/src/bootstrap/src/core/build_steps/compile.rs                                                                                 
@@ -1533,13 +1533,15 @@                                                                                                              
     /// must have been previously produced by the `stage - 1` builder.build                                                        
     /// compiler.                                                                                                                  
     fn run(self, builder: &Builder<'_>) -> Compiler {                                                                              
-        let target_compiler = self.target_compiler;                                                                                
+        let mut target_compiler = self.target_compiler;                                                                            
+        target_compiler.host = builder.config.build;                                                                                                                                              
                                                                                                                                    
         if target_compiler.stage == 0 {                                                                                            
             assert_eq!( 
                       builder.config.build, target_compiler.host,
                       "Cannot obtain compiler for non-native build triple at stage 0"
                       );
             // The stage 0 compiler for the build triple is always pre-built.
             return target_compiler;
         }
```

**Comment by onur-ozkan** [comment]
I will add another commit to thandle that.

**Comment by onur-ozkan** [comment]
> Do we still need the patch mentioned in [#115642 (comment)](https://github.com/rust-lang/rust/issues/115642#issuecomment-1878543949) along with the changes in #119619 PR?
> 
> I think the three file changes in the #119619 PR should be sufficient, right?

#119619 PR is enough, you don't need any other patches.

**Comment by Yashinde145** [comment]
> Here we have an issue for the remote-test tool problem #97669
> 
> It seems to occur randomly in QEMU environments from time to time. Retrying seems to fix the issue for each occurrence. I will try to identify the root cause in the coming week.

Hi @onur-ozkan, 
Did you get a chance to look at the `remote test-tool problem?`.

**Comment by onur-ozkan** [comment]
> Did you get a chance to look at the remote test-tool problem?

Not yet

**Comment by onur-ozkan** [comment]
Are you able to run any tests with remote-test or does it give the error instantly at the beginning?

**Comment by onur-ozkan** [comment]
I meant sending tests and running in the qemu environment

> I tested with the two commits and the errors of `using nightly flag and custom target specs` were not seen. Testing was completed for the test suites like ui, run-pass-valgrind, coverage, mir-opt, codegen, assembly, incremental.
> 
> But, after that while coping libs to qemu emulator(we run the test suite on qemu), following io::copy panic was observed:
> 
> ```
>  REMOTE copy libs to emulator (x86_64-poky-linux-gnu)                                                                          
>  Building stage0 tool remote-test-server (x86_64-unknown-linux-gnu -> x86_64-poky-linux-gnu)
>  
>                         
>  done pushing "/home/build-st/tmp/work/core2-64-poky-linux/rust/1.75.0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-poky-linux-gnu/lib/libcstr-d49e684941d617ce.so"                                   
>  done pushing "/home/build-st/tmp/work/core2-64-poky-linux/rust/1.75.0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-poky-linux-gnu/lib/libzerovec_derive-52a1b6c914833752.so"                         
>  done pushing "/home/build-st/tmp/work/core2-64-poky-linux/rust/1.75.0/rustc-1.75.0-src/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-poky-linux-gnu/lib/librustc_fluent_macro-4b4331439c3bd942.so"
>  
> thread 'main' panicked at src/tools/remote-test-client/src/main.rs:353:5: 
> io::copy(&mut file, dst) failed with Connection reset by peer (os error 104)                                                  
>  stack backtrace:                                                                                                              
>     0: rust_begin_unwind                                                                                                       
>               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/panicking.rs:597:5                            
>     1: core::panicking::panic_fmt                                                                                              
>               at /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/panicking.rs:72:14                           
>     2: remote_test_client::send                                                                                                
>     3: remote_test_client::main                                                                                                
>  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.                                       
>  Build completed unsuccessfully in 0:09:27
> ```

Is the error always the same, like this?

**Comment by Yashinde145** [comment]
Yes, the error always same.

**Comment by onur-ozkan** [comment]
Can you try this patch https://github.com/onur-ozkan/rust/commit/43183f7b70525a1ef7ec48e7ce7ddff82d9ff410 and tell me if it fixes the problem?

**Comment by Yashinde145** [comment]
The https://github.com/onur-ozkan/rust/commit/43183f7b70525a1ef7ec48e7ce7ddff82d9ff410 patch does not fix the problem. Still, the  same error is seen.

**Comment by onur-ozkan** [comment]
I was able to repreduce the `io::copy(&mut file, dst) failed with Connection reset by peer (os error 104)` error, but only by sending large files (and the patch I linked fixes it).

Unfortunately I don't know your use case, and I can't invest time on https://github.com/rust-lang/rust/issues/115642#issuecomment-1879634525 for debugging or trying to reproduce the error. If you can't debug and fix the problem you get, I suggest you to give more context about your use case (configurations, commands, etc.) under the https://github.com/rust-lang/rust/issues/97669 issue.

**Comment by Yashinde145** [comment]
I got some work around for this, 
by excluding `ui-fulldeps` test suite and `src/tools/replace-version-placeholder`, `src/tools/jsondoclint` tests and it works for `x86_64-poky-linux-gnu`.

But, when I check for other cross-compiled targets(`x86 32  bit, arm 32 & 64 bits and mips64`), the following lib failure is seen

```
error: linking with `cc` failed: exit status: 1

build-st/tmp/hosttools/ld:  build-st/tmp/work/core2-32-poky-linux/rust/1.74.1/rustc-1.74.1-src/build/x86_64-unknown-linux-gnu/stage1-std/i686-poky-linux-gnu-synthetic-miropt-abort/release/deps/std-55f86614dec482ec.3j457j23godxdmvr.rcgu.o: file class ELFCLASS32 incompatible with ELFCLASS64
build-st/tmp/hosttools/ld: final link failed: file in wrong format

collect2: error: ld returned 1 exit status
error: could not compile `std` (lib) due to previous error
```

I am investigating the problem and will let you know if there are any updates.

**Comment by onur-ozkan** [comment]
That seems to be related with MirOptPanicAbortSyntheticTarget tests https://github.com/rust-lang/rust/blob/f9c2421a2a6e34f3756900dd7d600704c08bfccb/src/bootstrap/src/core/build_steps/test.rs#L1530-L1534

**Comment by Yashinde145** [comment]
Yes, the log says the same and the error was seen after `MirOptPanicAbortSyntheticTarget ` test cases while building stage1 library artifacts.


Meanwhile, I came across the following and have a doubt,
As mentioned in https://github.com/rust-lang/rust/issues/40049#issuecomment-282082114, do we have to pass the  `-C linker=` flag to `RUSTFLAGS`  explicitly for the linker error in https://github.com/rust-lang/rust/issues/115642#issuecomment-1893148338?

Earlier it was working without passing the flag explicitly.

**Comment by onur-ozkan** [comment]
Earlier we didn't have those tests, so they might require passing `-C linker=`. You can try and see if that solves the problem.

**Comment by Yashinde145** [comment]
> That seems to be related with MirOptPanicAbortSyntheticTarget tests
> 
> https://github.com/rust-lang/rust/blob/f9c2421a2a6e34f3756900dd7d600704c08bfccb/src/bootstrap/src/core/build_steps/test.rs#L1530-L1534

Is there any way to handle this panic failure and make the tests run successfully?

**Comment by onur-ozkan** [comment]
You can add `--skip mir-opt` to your command to exclude mir-opt tests, then run `x test mir-opt` to run mir-opt tests without `MirOptPanicAbortSyntheticTarget`.
